crates/sui-framework/sources/safe.move

```diff
@@ -0,0 +1,214 @@
+        permissions_bitmap: vector<u64>,
+        next_nonce: u64,
+    }
+
+    ///
+    /// Allows the owner of the capability to take `amount` of coins from the box.
+    ///
+    /// @ownership: Owned
+    ///
+    struct Capability<phantom T> has store, key {
+        id: UID,
+        safe_id: ID,
+        nonce: u64,
+        // The amount that the user is able to transfer.
+        amount: u64,
+    }
+
+    //////////////////////////////////////////////////////
+    /// HELPER FUNCTIONS
+    //////////////////////////////////////////////////////
+    
+    /// Set an element in the bitmap
+    /// If one is set to be true, then the big-endian `index`-th bit in the bitmap will be set to 1.
+    /// Otherwise, the the big-endian `index`-th bit in the bitmap will be set to 0.
+    /// 
+    /// Note: This will panic if the index out of bounds.
+    public fun bitmap_set(bitmap: &mut vector<u64>, index: u64, one: bool) {
+        // Which chunk of bytes in the vector to set.
+        let vec_index = index / 64;
+        // Which bit in the u64 to set.
+        let bit_index = ((index % 64) as u8);
+
+        let byte_quartet = vector::borrow_mut(bitmap, vec_index);
+        let bitmask = (1 << (63 - bit_index));
+        if (one) {
+            *byte_quartet = *byte_quartet | bitmask;
+        } else {
+            bitmask = bitmask ^ 0xffffffff;
+            *byte_quartet = *byte_quartet & bitmask;
+        }
+    }
+
+    /// Expand a bitmap until the index matches
+    public fun bitmap_expand(bitmap: &mut vector<u64>, index: u64) {
+        let chunk_index = index / 64;
+        let num_chunks = vector::length(bitmap);
+        while (num_chunks <= chunk_index) {
+            vector::push_back(bitmap, 0);
+            num_chunks = num_chunks + 1;
+        }
+    }
+
+    /// Read an index in the bitmap and return whether the bit is equal to one.
+    /// 
+    /// Note: This will panic if the index out of bounds.
+    public fun bitmap_read(bitmap: &vector<u64>, index: u64): bool {
+        // Which u64 in the vector to read.
+        let vec_index = index / 64;
+        // Which bit in the u64 to read.
+        let bit_index = ((index % 64) as u8);
+
+        let byte_quartet = vector::borrow(bitmap, vec_index);
+        let bit_mask: u64 = (1 << (63 - bit_index));
+
+        return bit_mask & *byte_quartet != 0
+    }
+
+    /// Check that the capability has not yet been revoked by the owner.
+    fun check_capability_validity<T>(safe: &Safe<T>, capability: &Capability<T>) {
+        assert!(bitmap_read(&safe.permissions_bitmap, capability.nonce), 0);
+    }
+
+    /// Helper function to create a capability.
+    fun create_capability_<T>(safe: &mut Safe<T>, withdraw_amount: u64, ctx: &mut TxContext): Capability<T> {
+        let capability = Capability {
+            id: object::new(ctx),
+            safe_id: object::uid_to_inner(&safe.id),
+            nonce: safe.next_nonce,
+            amount: withdraw_amount,
```

Comment:
It is possible (though debatable whether it's desirable) for a `Safe` to have "overdraft protections"--it can remember how much it's promised via issuing `Capability`'s and make sure not to promise more than is in the `balance`.

Probably not be desired for all use-cases, but interesting to think about as an option.